rs/nns/governance/src/neuron_store.rs

```diff
@@ -1237,6 +1238,163 @@ impl NeuronStore {
+                            Some((original, current))
+                        } else {
+                            None
+                        }
+                    }
+                    Self::Borrowed(_) => None,
+                }
+            }
+        }
+
+        impl<'a> Deref for NeuronHandle<'a> {
+            type Target = Neuron;
+
+            fn deref(&self) -> &Neuron {
+                match self {
+                    Self::Owned { current, .. } => current,
+                    Self::Borrowed(neuron) => neuron,
+                }
+            }
+        }
+
+        // This records changes that need to be made to stable memory. That
+        // takes place later, because you can't modify while iterating.
+        let mut dirty_neurons = vec![];
+
+        with_stable_neuron_store(|stable_neuron_store| {
+            // Select stable memory neurons.
+            let stable_neuron = stable_neuron_store
+                .range_neurons_sections(neuron_ids, neuron_sections)
+                .map(|neuron| {
+                    // With the help of Cow, it might be possible to avoid this clone...
+                    let original = neuron.clone();
+                    NeuronHandle::Owned {
+                        original,
+                        current: neuron,
+                    }
+                });
+
+            // Select heap neurons.
+            let neuron_ids = {
+                fn to_u64(bound: Bound<&NeuronId>) -> Bound<u64> {
+                    match bound {
+                        Bound::Unbounded => Bound::Unbounded,
+                        Bound::Included(neuron_id) => Bound::Included(neuron_id.id),
+                        Bound::Excluded(neuron_id) => Bound::Excluded(neuron_id.id),
+                    }
+                }
+
+                let start = to_u64(neuron_ids.start_bound());
+                let end = to_u64(neuron_ids.end_bound());
+
+                (start, end)
+            };
+            let heap_neuron = self
+                .heap_neurons
+                .range_mut(neuron_ids)
+                .map(|(_, neuron)| NeuronHandle::Borrowed(neuron));
+
+            // Merge neurons in heap + stable memory.
+            let neurons = stable_neuron.merge_by(heap_neuron, |neuron_1, neuron_2| {
+                neuron_1.id() <= neuron_2.id()
+            });
+
+            // Scan neurons.
+            for mut neuron in neurons {
+                let carry_on = fun(neuron.mut_ref());
+
+                if let Some((original, current)) = neuron.take_change() {
+                    dirty_neurons.push((original, current));
+                }
+
+                if !carry_on {
+                    break;
+                }
+            }
+        });
+
+        // Write changes to stable memory.
+        with_stable_neuron_store_mut(|stable_neuron_store| {
+            for (original, current) in dirty_neurons {
+                let update = stable_neuron_store.update(&original, current);
```

Comment:
Jason pointed out - this isn't updating the neuron indices either.

Is there a way we can use with_neuron_mut instead?

Reply:
Done.